Python/chapter01/1.3 - URLify/miguel_1.3_sol.py

```diff
@@ -0,0 +1,87 @@
+"""
+Python version 3.7.0
+1.3 - URLify
+Write a method to replace all spaces in a string with "%20".
+The following only applies to C/C++: "You may assume that the string has sufficient
+space at the end to hold the additional characters, and that you are given the "true" length of the string"
+(Note: If implementing in Java, please use a character array so that you can perform this operation in place)
+Since we are using python, no need to use true length.  I will have two functions, one with
+true_length and the other without.
+"""
+import unittest
+from typing import Callable
+
+
+def urlify(s: str, true_length: int) -> str:
+    """
+    Given a string and it's "true" length, this function will return a new string
+    that replaces all of the spaces of the input string with '%20'.
+    Precondition(s):
+    - length of s <= true_length
+    :param s: the original string to 'urlify'
+    :param true_length: since s may have additional characters, we focus on true_length instead of actual s length
+    :return: string with each space from s replaced with '%20'
+    """
+    # Below link goes over string concat efficiency in python (I will use method 4
+    # https://waymoot.org/home/python_string/
+    output = []
+    for c in s[0:true_length]:
```

Comment:
optional: you already have a solution in the Pythonic way (urlify_no_true_length), but if you want to do the exercise as the author of the book intended, try solving it in under the following constraints:

```python3
result = ['\x00'] * true_length
for i, c in s:
  result[i] = c
# your code. you are only allowed to allocate O(1) extra memory.
return ''.join(result).rstrip('\x00')
```

Reply:
Gave it an attempt.  My attempt was not as clean as the pythonic way, but it was fun to try it out.  It somehow felt odd to code in python as if it was java or c/c++ lol.  But I also feel that it was good practice.

I changed the result length from true_length to len(s) * 2, since for each character up to true_length in s, the size of result will increase by 2 characters.